### PR TITLE
Avoid part ID references in Refits and elsewhere

### PR DESCRIPTION
**Pull request by sixlettervariables**
This PR is nasty and is not going to be fun to review...but at this point I am happy with just about all of the part ID issues I have been made aware of in the past. Basically: we're not going to reference parts by ID anywhere other than the CPNX file. Everywhere else should be using an actual reference to a Part object. That's about as close to avoiding part ID issues as you can get.

This is a draft because I'm not 100% on the refit code, but close.

### CODE REVIEW

**Comment by sixlettervariables** [comment]
This is as comfortable as I am with a refit change without more testing in the wild.

This also stages some changes to AmmoBin for later, where we pull that logic out and into the Campaign (or hopefully somewhere more relevant like a Warehouse class).

**Comment by sixlettervariables** [comment]
Known issues:
- Custom refits do not tag purchased ammo as "reserved for refit" as opposed to refit kits (reason unknown), which means you **may** have to purchase ammo again if during the refit you use it.
- Cancelled refits may give you ammo bins.

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 1]
```diff
@@ -1634,7 +1634,8 @@ public void addPart(Part p, int transitDays) {
             p.setId(-1);
             return;
         }
-        if ((null == p.getUnit()) && !p.hasParentPart()) {
+        if ((null == p.getUnit()) && !p.hasParentPart()
+            && !p.isReservedForRefit() && !p.isReservedForReplacement()) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:1638]:
```suggestion
                && !p.isReservedForRefit() && !p.isReservedForReplacement()) {
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 2]
```diff
@@ -1746,7 +1747,8 @@ public void addPartWithoutId(Part p) {
         // go ahead and check for existing parts because some version weren't
         // properly collecting parts
         Part mergedWith = null;
-        if (!(p instanceof MissingPart) && null == p.getUnitId()) {
+        if (!(p instanceof MissingPart) && (null == p.getUnitId())
+            && !p.isReservedForRefit() && !p.isReservedForReplacement()) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:1751]:
```suggestion
                && !p.isReservedForRefit() && !p.isReservedForReplacement()) {
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 3]
```diff
@@ -1773,39 +1775,19 @@ public void addPartWithoutId(Part p) {
             parts.put(p.getId(), p);
             MekHQ.triggerEvent(new PartNewEvent(p));
         } else {
-            // Go through each unit and its refits to see if the new armor ID should be updated
+            // Go through each unit and its refits to see if the new armorshould be updated
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:1778]:
```suggestion
            // Go through each unit and its refits to see if the new armor should be updated
```

File: MekHQ/src/mekhq/campaign/Campaign.java [hunk 4]
```diff
@@ -1773,39 +1775,19 @@ public void addPartWithoutId(Part p) {
...
+                    Refit r = u.getRefit();
+                    // If there is a refit and this part matches the new armor, update the armor entry
+                    if (null != r) {
+                        if (r.getNewArmorSupplies() == p) {
+                            MekHQ.getLogger().info(
+                                String.format("%s (%d) was merged with %s (%d) used in a refit for %s", p.getName(),
                                     p.getId(), mergedWith.getName(), mergedWith.getId(), u.getName()));
-                                ids.set(ii, mergedWith.getId());
-                            }
+                            Armor mergedArmor = (Armor)mergedWith;
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/Campaign.java:1789]:
```suggestion
                            Armor mergedArmor = (Armor) mergedWith;
```

File: MekHQ/src/mekhq/campaign/parts/MissingPart.java [hunk 5]
```diff
@@ -161,12 +161,10 @@ public Part findReplacement(boolean refit) {
...
-		if(replacementId > -1) {
-			bestPart = campaign.getPart(replacementId);
-			if(null != bestPart) {
-				return bestPart;
-			}
-		}
+		if (hasReplacementPart()) {
+			return getReplacementPart();
+        }
+
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/MissingPart.java:167]: Tab vs Space formatting

File: MekHQ/src/mekhq/campaign/parts/MissingPart.java [hunk 6]
```diff
@@ -347,11 +345,13 @@ public void writeToXmlBegin(PrintWriter pw1, int indent) {
...
-		pw1.println(MekHqXmlUtil.indentStr(indent+1)
-				+"<replacementId>"
-				+replacementId
-				+"</replacementId>");
+                +"</daysToWait>");
+        if (hasReplacementPart()) {
+            pw1.println(MekHqXmlUtil.indentStr(indent+1)
+                    +"<replacementId>"
+                    + getReplacementPart().getId()
+                    +"</replacementId>");
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/MissingPart.java:353]:
```suggestion
            MekHqXmlUtil.writeSimpleXmlTag(pw1, indent + 1, "replacementId", getReplacementPart().getId());
```

File: MekHQ/src/mekhq/campaign/parts/MissingPart.java [hunk 7]
```diff
@@ -402,29 +402,29 @@ public void reservePart() {
...
 			}
 		}
 	}
 
 	@Override
 	public void cancelReservation() {
 		Part replacement = findReplacement(false);
-		if(replacementId > -1 && null != replacement) {
-			replacementId = -1;
+		if (hasReplacementPart() && null != replacement) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/MissingPart.java:422]:
```suggestion
		if (hasReplacementPart() && (null != replacement)) {
```

File: MekHQ/src/mekhq/campaign/parts/Part.java [hunk 8]
```diff
@@ -1783,6 +1823,16 @@ public SimpleTechLevel getStaticTechLevel() {
     }
 
     public void fixupPartReferences(Map<Integer, Part> knownParts) {
+        if (replacementPart instanceof PartRef) {
+            int id = replacementPart.getId();
+            replacementPart = knownParts.get(id);
+            if ((replacementPart == null) && (id > 0)) {
+                MekHQ.getLogger().error(PartRef.class,
+                    String.format("Part %d ('%s') references missing replacement part %d",
+                        getId(), getName(), id));
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Part.java:1832]:
```suggestion
                MekHQ.getLogger().error(
                    String.format("Part %d ('%s') references missing replacement part %d",
                        getId(), getName(), id));
```

**Review by Windchild292 — COMMENTED** [review]
First a few suggestions

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 9]
```diff
@@ -508,15 +503,15 @@ public void calculate() {
                 Part replacement = ((MissingPart)nPart).findReplacement(true);
                 //check quantity
                 //TODO: the one weakness here is that we will not pick up damaged parts
-                if (null != replacement && null == partQuantity.get(replacement.getId())) {
-                    partQuantity.put(replacement.getId(), replacement.getQuantity());
+                if (null != replacement && null == partQuantity.get(replacement)) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:506]:
```suggestion
                if ((null != replacement) && (null == partQuantity.get(replacement))) {
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 10]
```diff
@@ -508,15 +503,15 @@ public void calculate() {
...
+                    partQuantity.put(replacement, replacement.getQuantity());
                 }
-                if (null != replacement && partQuantity.get(replacement.getId()) > 0) {
-                    newUnitParts.add(replacement.getId());
+                if (null != replacement && partQuantity.get(replacement) > 0) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:509]:
```suggestion

                if ((null != replacement) && (partQuantity.get(replacement) > 0)) {
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 11]
```diff
@@ -508,15 +503,15 @@ public void calculate() {
...
+                    newUnitParts.add(replacement);
                     //adjust quantity
-                    partQuantity.put(replacement.getId(), partQuantity.get(replacement.getId())-1);
+                    partQuantity.put(replacement, partQuantity.get(replacement)-1);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:512]:
```suggestion
                    partQuantity.put(replacement, partQuantity.get(replacement) - 1);
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 12]
```diff
@@ -554,15 +549,15 @@ public void calculate() {
                     Part replacement = mab.findReplacement(true);
                     //check quantity
                     //TODO: the one weakness here is that we will not pick up damaged parts
-                    if (null != replacement && null == partQuantity.get(replacement.getId())) {
-                        partQuantity.put(replacement.getId(), replacement.getQuantity());
+                    if (null != replacement && null == partQuantity.get(replacement)) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:552]:
```suggestion
                    if ((null != replacement) && (null == partQuantity.get(replacement))) {
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 13]
```diff
@@ -554,15 +549,15 @@ public void calculate() {
...
+                        partQuantity.put(replacement, replacement.getQuantity());
                     }
-                    if (null != replacement && partQuantity.get(replacement.getId()) > 0) {
-                        newUnitParts.add(replacement.getId());
+                    if (null != replacement && partQuantity.get(replacement) > 0) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:555]:
```suggestion

                    if ((null != replacement) && (partQuantity.get(replacement) > 0)) {
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 14]
```diff
@@ -1214,7 +1224,8 @@ public boolean acquireParts() {
             return false;
         }
 
-        return shoppingList.size() == 0 && !missingAmmo && (null == newArmorSupplies || (armorNeeded - newArmorSupplies.getAmount()) <= 0);
+        return shoppingList.size() == 0
+            && (null == newArmorSupplies || (armorNeeded - newArmorSupplies.getAmount()) <= 0);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:1228]:
```suggestion
            && ((null == newArmorSupplies) || (armorNeeded - newArmorSupplies.getAmount()) <= 0);
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 15]
```diff
@@ -1276,13 +1275,20 @@ private void updateRefitClass(int rClass) {
...
-            Part part = oldUnit.getCampaign().getPart(pid);
-            if (null == part) {
-                MekHQ.getLogger().error(this, "part with id " + pid + " not found for refit of " + getDesc());
-                continue;
+
+        // Return our reserved ammo back to the campaign
+        for (Part part : newUnitParts) {
+            if (part instanceof AmmoStorage) {
+                // merge back into the campaign
+                AmmoStorage ammoStorage = (AmmoStorage)part;
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:1283]:
```suggestion
                AmmoStorage ammoStorage = (AmmoStorage) part;
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 16]
```diff
@@ -1276,13 +1275,20 @@ private void updateRefitClass(int rClass) {
...
+                AmmoBin.changeAmountAvailable(campaign, ammoStorage.getShots(), (AmmoType)ammoStorage.getType());
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:1284]:
```suggestion
                AmmoBin.changeAmountAvailable(campaign, ammoStorage.getShots(), (AmmoType) ammoStorage.getType());
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 17]
```diff
@@ -1465,6 +1458,12 @@ else if ((part instanceof AeroHeatSink) && (newEntity instanceof Aero) && !part.
                     oldUnit.getCampaign().removePart(part);
                     continue;
                 }
+            } else if (part instanceof AmmoStorage) {
+                // merge back into the campaign before completing the refit
+                AmmoStorage ammoStorage = (AmmoStorage)part;
+                AmmoBin.changeAmountAvailable(campaign, ammoStorage.getShots(), (AmmoType)ammoStorage.getType());
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:1464]:
```suggestion
                AmmoStorage ammoStorage = (AmmoStorage) part;
                AmmoBin.changeAmountAvailable(campaign, ammoStorage.getShots(), (AmmoType) ammoStorage.getType());
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 18]
```diff
@@ -2589,4 +2581,166 @@ public boolean isIntroducedBy(int year, boolean clan, int techFaction) {
     public boolean isExtinctIn(int year, boolean clan, int techFaction) {
         return isExtinct(year, clan, techFaction);
     }
+
+    @Override
+    public void fixupPartReferences(Map<Integer, Part> knownParts) {
+        if (newArmorSupplies instanceof RefitArmorRef) {
+            Part realPart = knownParts.get(newArmorSupplies.getId());
+            if (realPart instanceof Armor) {
+                newArmorSupplies = (Armor)realPart;
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:2590]:
```suggestion
                newArmorSupplies = (Armor) realPart;
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 19]
```diff
@@ -2589,4 +2581,166 @@ public boolean isIntroducedBy(int year, boolean clan, int techFaction) {
...
+            } else {
+                MekHQ.getLogger().error(RefitArmorRef.class,
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:2592]:
```suggestion
                MekHQ.getLogger().error(
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 20]
```diff
@@ -2589,4 +2581,166 @@ public boolean isIntroducedBy(int year, boolean clan, int techFaction) {
...
+        }
+
+        for (int ii = oldUnitParts.size() - 1; ii >= 0; --ii) {
+            Part part = oldUnitParts.get(ii);
+            if (part instanceof RefitPartRef) {
+                Part realPart = knownParts.get(part.getId());
+                if (realPart != null) {
+                    oldUnitParts.set(ii, realPart);
+                } else if (part.getId() > 0) {
+                    MekHQ.getLogger().error(RefitPartRef.class,
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:2606]:
```suggestion
                    MekHQ.getLogger().error(
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 21]
```diff
@@ -2589,4 +2581,166 @@ public boolean isIntroducedBy(int year, boolean clan, int techFaction) {
...
+        }
+
+        for (int ii = newUnitParts.size() - 1; ii >= 0; --ii) {
+            Part part = newUnitParts.get(ii);
+            if (part instanceof RefitPartRef) {
+                Part realPart = knownParts.get(part.getId());
+                if (realPart != null) {
+                    newUnitParts.set(ii, realPart);
+                } else if (part.getId() > 0) {
+                    MekHQ.getLogger().error(RefitPartRef.class,
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:2621]:
```suggestion
                    MekHQ.getLogger().error(
```

File: MekHQ/src/mekhq/campaign/parts/Refit.java [hunk 22]
```diff
@@ -2589,4 +2581,166 @@ public boolean isIntroducedBy(int year, boolean clan, int techFaction) {
...
+        Iterator<Part> it = lcBinsToChange.iterator();
+        while (it.hasNext()) {
+            Part part = it.next();
+            if (part instanceof RefitPartRef) {
+                Part realPart = knownParts.get(part.getId());
+                it.remove();
+                if (realPart != null) {
+                    realParts.add(realPart);
+                } else {
+                    MekHQ.getLogger().error(RefitPartRef.class,
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/parts/Refit.java:2639]:
```suggestion
                    MekHQ.getLogger().error(
```

**Review by Windchild292 — COMMENTED** [review]
More formatting comments

**Review by Windchild292 — APPROVED** [review]
Having now gone over this thrice over two days, I think this looks good.

**Pull request merged**